### PR TITLE
HD5: Label balancing, hd5 read speed improvement, refactoring

### PR DESCRIPTION
**Pull request by torbjorgen**
- HD5 generator, TF Dataset: Label sampling
- HD5 generator: Rework multiclass label signature
- HD5 generator,TF data.Dataset: Reloadable generator
- HD5 data generators: Support for data labels, bugfix
- HD5 generator: Drop loop forever argument
- HD5 data generator: Load data into RAM for performance.

### CODE REVIEW

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 1]
```diff
@@ -12,40 +12,49 @@
...
         self._data_length = int
         self._replace_nan_floats_with = replace_nan_floats_with
 
         self._group_name = group_name if group_name else list(self._hd5_file.keys())[0]
         _LOGGER.info(f'Generating data from group \'{group_name}\'')
         self._group_name: str = group_name if group_name else list(self._hd5_file.keys())[0]
         self._group: h5py.Group = self._hd5_file[self._group_name]
+        if load_data_into_ram:
+            # Load all data into RAM as numpy arrays.
+            in_ram_group: Dict[str, np.ndarray] = dict()  # Mock the Group dictionary API by using a dict
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
            in_ram_group: Dict[str, np.ndarray] = {}  # Mock the Group dictionary API by using a dict
```

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 2]
```diff
@@ -12,40 +12,49 @@
...
+            for dataset_name in self._group.keys():
+                in_ram_group.update({dataset_name: self._group[dataset_name][:]})
+            self._group = in_ram_group
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hint

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 3]
```diff
@@ -95,6 +100,7 @@ def _replace_nan_values(self, sample: Union[bytes, float]) -> Union[bytes, float
         raise NotImplementedError(f'Replacing NaN values for dtype {type(sample)} not implemented.')
 
     def _assemble_output_vector(self,
+                                idx,
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hint

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 4]
```diff
@@ -154,31 +161,30 @@ def _expand_categorical_label_1d_to_2d(label: float) -> Tuple[float, float]:
...
         Yields data from HD5 data set.
         :return:
             1. Tuple of data (tensor0, tensor1, ...) OR
             2. Tuple of tuples ((tensor0, tensor1, ...), (label, ))
         """
-        while True:
-            x = self._assemble_output_vector(self._output_tensor_format)
+        idx = 0
+        while idx < self._data_length:
+            x = self._assemble_output_vector(idx=idx,
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]:
```suggestion
            output_vector = self._assemble_output_vector(idx=idx,
```
Type hint?

File: src/rdds/lib/hdf5/hd5_data_generator.py [hunk 5]
```diff
@@ -154,31 +161,30 @@ def _expand_categorical_label_1d_to_2d(label: float) -> Tuple[float, float]:
...
+                                             output_tensor_format=self._output_tensor_format)
             if self._label is not None:
-                label = self._assemble_output_vector([self._label])
+                label = self._assemble_output_vector(idx=idx,
```
Comment (henrikstranneheim) [inline src/rdds/lib/hdf5/hd5_data_generator.py]: Add type hint

File: src/rdds/lib/tf/dataset_rejection_resample.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+from rdds.lib.determinism import SEED
+
+
+def rejection_resample(dataset: tf.data.Dataset,
+                       desired_class_ratio: List[float],
+                       seed: int = SEED) -> tf.data.Dataset:
+    """
+
+    :param dataset:
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/dataset_rejection_resample.py]: Update

File: src/rdds/lib/tf/dataset_rejection_resample.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    :return:
+    """
+
+    def class_mapping_fn(data: Tuple[tf.Tensor],
+                         labels: Tuple[tf.Tensor]):
+        labels: tf.Tensor = labels[0]  # Unpack tuple
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/dataset_rejection_resample.py:27]:
```suggestion
        labels: tf.Tensor = labels[0]
```

File: src/rdds/lib/tf/dataset_rejection_resample.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        print('labels', labels)
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/dataset_rejection_resample.py]: Use log instead

File: src/rdds/lib/tf/dataset_rejection_resample.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                       desired_class_ratio: List[float],
+                       seed: int = SEED) -> tf.data.Dataset:
+    """
+
+    :param dataset:
+    :return:
+    """
+
+    def class_mapping_fn(data: Tuple[tf.Tensor],
+                         labels: Tuple[tf.Tensor]):
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/dataset_rejection_resample.py]: Add return hint

File: src/rdds/lib/tf/dataset_rejection_resample.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        labels: tf.Tensor = labels[0]  # Unpack tuple
+        print('labels', labels)
+        return tf.cast(tf.math.equal(labels[0], 1), dtype=tf.int32)
+
+    if not len(desired_class_ratio) == 2:
```
Comment (henrikstranneheim) [inline src/rdds/lib/tf/dataset_rejection_resample.py]: Why 2 can you explain with a constant?

File: src/tests/fixtures/hdf5_data.py [hunk 11]
```diff
@@ -39,3 +41,15 @@ def hd5_file_path_with_nans(hd5_file):
     hd5_file.flush()
     hd5_file.close()
     yield hd5_file_path
+
+@pytest.fixture
```
Comment (henrikstranneheim) [inline src/tests/fixtures/hdf5_data.py:45]: I think you should move all fixtures to conftests modules before they become to many, but in a dedicated PR.

File: src/tests/fixtures/hdf5_data.py [hunk 12]
```diff
@@ -39,3 +41,15 @@ def hd5_file_path_with_nans(hd5_file):
...
+def hd5_file_path_with_categorical_labels(hd5_file) -> str:
```
Comment (henrikstranneheim) [inline src/tests/fixtures/hdf5_data.py]: Add type hint and doc str

File: src/tests/hdf5/test_generator.py [hunk 13]
```diff
@@ -29,23 +28,21 @@ def test_hd5_data_generator(hd5_file):
         data_iter.__next__()
 
 
-def test_hd5_data_generator_repeat(hd5_file):
+def test_hd5_data_generator_nr_samples(hd5_file):
     """
-    Test for repeating data.
+    Test for data length.
```
Comment (henrikstranneheim) [inline src/tests/hdf5/test_generator.py]: Could this be clarified?

File: src/tests/tf/test_dataset_rejection_resample.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+from rdds.lib.hdf5 import Hd5DataGenerator
+
+# Test fixtures
+from tests.tf.test_data_generator import output_signature
+
+from rdds.lib.tf import rejection_resample
+
+
+@pt.fixture
+def hd5_data_generator(hd5_file_path_with_categorical_labels) -> Hd5DataGenerator:
```
Comment (henrikstranneheim) [inline src/tests/tf/test_dataset_rejection_resample.py]: Add type hint

File: src/tests/tf/test_dataset_rejection_resample.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+                            label='label')
+
+
+@pt.mark.parametrize("expected_negative_ratio, expected_positive_ratio",
+                     [(1.00, 0.00),
+                      (0.25, 0.75),
+                      (0.50, 0.50),
+                      (0.75, 0.25),
+                      (0.00, 1.00)])
+def test_balanced_sampling(expected_negative_ratio,
```
Comment (henrikstranneheim) [inline src/tests/tf/test_dataset_rejection_resample.py]: Type hints

File: src/tests/tf/test_dataset_rejection_resample.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+                           expected_positive_ratio,
+                           hd5_data_generator,
+                           output_signature):
+    """
+    Test for dataset to provide 50% of samples from the two classes
+    """
+    # GIVEN a dataset
+    data_set: tf.data.Dataset = \
```
Comment (henrikstranneheim) [inline src/tests/tf/test_dataset_rejection_resample.py:36]: Do you use black?

**Review by henrikstranneheim — APPROVED** [review]
👍

File: src/tests/tf/test_dataset_rejection_resample.py (hunk 16 above)
Comment (torbjorgen) [reply src/tests/tf/test_dataset_rejection_resample.py:36]: No I do not. I'll give it a try @henrikstranneheim :+1: https://github.com/Clinical-Genomics/rdds/issues/53

**Comment by torbjorgen** [comment]
* Updated according to @henrikstranneheim  suggestions.

**Pull request merged**